Lib/test/test_pickle.py

```diff
@@ -570,6 +570,87 @@ def test_multiprocessing_exceptions(self):
                                  ('multiprocessing.context', name))
 
 
+class C_with_reduce_ex:
+    def __reduce_ex__(self, proto):
+        return object.__reduce_ex__(self, proto)
+
+class C_None_reduce_ex:
+    __reduce_ex__ = None
+
+class C_None_reduce:
+    __reduce__ = None
+
+class C_None_setstate:
+    def __getstate__(self):
+        return 1
+
+    __setstate__ = None
+
+class C_setstate:
+    def __init__(self):
+        self.value = 0
+
+    def __getstate__(self):
+        return self.value
+
+    def __setstate__(self, state):
+        print(state)
+        self.value = state + 1
+
+class PickleReductionMethodsTests(unittest.TestCase):
+    # see gh-93627
+
+    def test_pickle_invalid_reduction_method(self):
+        c = C_None_reduce_ex()
+        with self.assertRaises(TypeError):
+            pickle.dumps(c)
```

Comment:
What implementation does it test?

The tests in this module is organized in special way. There are abstract classes which test with the specified implementation specified by class attributes, and there are concrete subclasses which specialize tests for Python and C implementations. It would be better to rewrite your tests in conformance with existing style.

Reply:
@serhiy-storchaka  I moved the tests to be part of `AbstractPickleTests`, `AbstractPicklerUnpicklerObjectTests` and `AbstractDispatchTableTests`. There are included in concrete tests both inside and outside the `has_c_implementation` section of `test_pickle.py`.